### PR TITLE
UI: Checks the clusters version to ensure it's up to date.

### PR DESCRIPTION
**Pull request by maxlang**
- Shows a warning if the cluster is out of date
- Shows a warning if the node versions are mismatched

screen shot 2016-03-25 at 5 09 21 am

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5569)

### CODE REVIEW

File: ui/ts/components/helpusprompt.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+      // set modal container class
+      if (ctrl.stage === Stage.ModalClose || ctrl.stage === Stage.ThanksClose) {
+        containerClass = ".close";
+      }
+
+      if (ctrl.show) {
+          // Banner
+        switch (ctrl.stage) {
+          case Stage.Banner:
+            return m.component(Components.Banner, {
```
Comment (maxlang) [inline ui/ts/components/helpusprompt.ts]: This case in the switch statement was changed to use the new Banner component.

File: ui/ts/components/helpusprompt.ts [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+          case Stage.ModalStart:
+          case Stage.Modal:
+          case Stage.ModalClose:
+            return m.component(Components.Modal, {
+              containerClass: containerClass,
+              modalClass: modalClass,
+              title: "Help Cockroach Labs",
+              content: [
+                m(".intro", `CockroachDB is in beta, and we're working diligently to make it better.
+                Opt in to share basic, anonymous usage statistics.`),
```
Comment (maxlang) [inline ui/ts/components/helpusprompt.ts]: The wording here was changed to skip mentioning checking for versions.

**Comment by maxlang** [comment]
```ts/pages/helpusprompt.ts``` was moved to ```ts/components/helpusprompt.ts``` but since the indentation was changed it doesn't look like a rename in the PR. I've called out the only changes.

**Comment by maxlang** [comment]
cc @cuongdo 
cc @spencerkimball this will hit the registration server (though it doesn't work right now because of https://github.com/cockroachlabs/registration/issues/12)

**Comment by maxlang** [comment]
Example outdated banner:

screen shot 2016-03-25 at 3 09 36 pm

**Comment by mrtracy** [comment]
Reviewed 10 of 20 files at r1, 2 of 2 files at r2.
Review status: 9 of 19 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[ui/ts/models/cockroachlabs.ts, line 60 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDj7OtiTRJ8E-SIyGeV:-KDj7Otj3n54QCLskOtS:-1558834598) ([raw file](https://github.com/cockroachdb/cockroach/blob/836a8930b7a5df0d7190adeec069eda6ddf41930/ui/ts/models/cockroachlabs.ts#L60)):*
Do we need this here? I am not clear on how this work overlaps with the work @dt has done to upload usage information.

If we're only reading information (which I believe to be the case), then the cluster ID is not necessary and should be omitted here.

---

*[ui/ts/models/cockroachlabs.ts, line 62 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDj9I0QBCdJiStEiiAB:-KDj9I0QBCdJiStEiiAC:1000319317) ([raw file](https://github.com/cockroachdb/cockroach/blob/836a8930b7a5df0d7190adeec069eda6ddf41930/ui/ts/models/cockroachlabs.ts#L62)):*
This is why we need a model container, it's really gotten out of hand.

---

*[ui/ts/models/cockroachlabs.ts, line 68 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDk-uaNy4rERtaFVxTr:-KDk-uaNy4rERtaFVxTs:-179857292) ([raw file](https://github.com/cockroachdb/cockroach/blob/815bcdb324865804d7f4a4ac5aa45649accfe442/ui/ts/models/cockroachlabs.ts#L68)):*
This could just be moved inside of the `versionCheck()` method I think;

---

*[ui/ts/models/cockroachlabs.ts, line 76 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDk-HSQNyEQu4z5q6yJ:-KDk-HSQNyEQu4z5q6yK:682457202) ([raw file](https://github.com/cockroachdb/cockroach/blob/815bcdb324865804d7f4a4ac5aa45649accfe442/ui/ts/models/cockroachlabs.ts#L76)):*
This comment needs more detail; it always checks cluster mismatch, but only checks against cockroach's official version once a day.

---

*[ui/ts/models/cockroachlabs.ts, line 83 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDk-meN1Z4M9C4P7bSK:-KDk-meN1Z4M9C4P7bSL:751492840) ([raw file](https://github.com/cockroachdb/cockroach/blob/815bcdb324865804d7f4a4ac5aa45649accfe442/ui/ts/models/cockroachlabs.ts#L83)):*
Maybe a comment here; this is what eventually resolves the outstanding promise from version check.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569)*

**Comment by dt** [comment]
Review status: 9 of 19 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[ui/ts/models/cockroachlabs.ts, line 60 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDj7OtiTRJ8E-SIyGeV:-KDkBRRSxvZFQieqNOQZ:-177701432) ([raw file](https://github.com/cockroachdb/cockroach/blob/836a8930b7a5df0d7190adeec069eda6ddf41930/ui/ts/models/cockroachlabs.ts#L60)):*
Eh, I think we have to assume that most clusters won't have direct internet access, so while the server-side reporting will make a best-effort, I expect in most real deployments, the UI will be our best bet for actually checking for new versions.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569)*

**Comment by maxlang** [comment]
cc @mberhault 

This will start hitting the registration server when it's merged.

**Comment by maxlang** [comment]
Review status: 9 of 19 files reviewed at latest revision, 7 unresolved discussions.

---

*[ui/ts/models/cockroachlabs.ts, line 60 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDj7OtiTRJ8E-SIyGeV:-KDk93sTEPh6aMJibE4-:-798382905) ([raw file](https://github.com/cockroachdb/cockroach/blob/836a8930b7a5df0d7190adeec069eda6ddf41930/ui/ts/models/cockroachlabs.ts#L60)):*
It's not strictly necessary but will give us a sense of the number of unique clusters.

---

*[ui/ts/models/cockroachlabs.ts, line 62 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDj9I0QBCdJiStEiiAB:-KDkCGNBJPJTLNVxLZiJ:-480062378) ([raw file](https://github.com/cockroachdb/cockroach/blob/836a8930b7a5df0d7190adeec069eda6ddf41930/ui/ts/models/cockroachlabs.ts#L62)):*
Yeah.

---

*[ui/ts/models/cockroachlabs.ts, line 68 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDk-uaNy4rERtaFVxTr:-KDk9NPxURuaLp2ijitJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/815bcdb324865804d7f4a4ac5aa45649accfe442/ui/ts/models/cockroachlabs.ts#L68)):*
Done.

---

*[ui/ts/models/cockroachlabs.ts, line 76 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDk-HSQNyEQu4z5q6yJ:-KDkKf99_sSS6-JV2EKW:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/815bcdb324865804d7f4a4ac5aa45649accfe442/ui/ts/models/cockroachlabs.ts#L76)):*
Done

---

*[ui/ts/models/cockroachlabs.ts, line 83 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDk-meN1Z4M9C4P7bSK:-KDkKvHThcXp5TamdqBg:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/815bcdb324865804d7f4a4ac5aa45649accfe442/ui/ts/models/cockroachlabs.ts#L83)):*
Done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569)*

**Comment by maxlang** [comment]
I changed the behavior a little bit. It checks the version until you dismiss the banner. Then it waits a day before checking again. We should probably store the latest seen version/issue so that it only alerts you once per issue once dismissed, but that can be in a separate change.

---

Review status: 9 of 19 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-:-KDkLNJmFFnTvPl9vjDQ:863807115)*

**Comment by cuongdo** [comment]
Gave this a quick review. Since my TypeScript/Mithril knowledge isn't great, I can help more by testing this once it's merged.

---

Reviewed 4 of 20 files at r1, 2 of 5 files at r3, 2 of 3 files at r4.
Review status: 17 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[ui/ts/models/api.ts, line 115 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDkhMRpXIQn3mV2TZFy:-KDkhMRpXIQn3mV2TZFz:-1100501658) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b1b6abd0a9e22d78032aa6c00d86469bbf38e9/ui/ts/models/api.ts#L115)):*
Is there a named constant somewhere that can be used instead of `5`?

---

*[ui/ts/models/cockroachlabs.ts, line 11 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDkiafMdA2FP_86cKlG:-KDkiafMdA2FP_86cKlH:495913493) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b1b6abd0a9e22d78032aa6c00d86469bbf38e9/ui/ts/models/cockroachlabs.ts#L11)):*
Don't leave us hanging!

---

*[ui/ts/models/cockroachlabs.ts, line 64 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDkjjIn7QH3Shs5am_b:-KDkjjIookFZFpHkALh5:400127471) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b1b6abd0a9e22d78032aa6c00d86469bbf38e9/ui/ts/models/cockroachlabs.ts#L64)):*
what's the `1.` prefix mean?

---

*[ui/ts/models/cockroachlabs.ts, line 105 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDkkRN7VYvMauQuYB7G:-KDkkRN7VYvMauQuYB7H:378136059) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b1b6abd0a9e22d78032aa6c00d86469bbf38e9/ui/ts/models/cockroachlabs.ts#L105)):*
what is `localhost:8084`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-:-KDkla_Uh9Fum3ctps3Z:875777167)*

**Comment by maxlang** [comment]
Thanks!

---

Review status: 15 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[ui/ts/models/api.ts, line 115 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDkhMRpXIQn3mV2TZFy:-KDohgQBLq3iXgsINsQc:1445803139) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b1b6abd0a9e22d78032aa6c00d86469bbf38e9/ui/ts/models/api.ts#L115)):*
Actually this doesn't matter anymore. Ever since your last change it will just return ```{}``` if no key is found (I think?). I'll remove this catch.

---

*[ui/ts/models/cockroachlabs.ts, line 11 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDkiafMdA2FP_86cKlG:-KDohkmZj4BXuMvogmm1:-854027645) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b1b6abd0a9e22d78032aa6c00d86469bbf38e9/ui/ts/models/cockroachlabs.ts#L11)):*
Haha and registration.

---

*[ui/ts/models/cockroachlabs.ts, line 64 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDkjjIn7QH3Shs5am_b:-KDogdgk4ijnuZgOm0rv:81831918) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b1b6abd0a9e22d78032aa6c00d86469bbf38e9/ui/ts/models/cockroachlabs.ts#L64)):*
It just refers to the 2nd element in the array. So this is basically ```data[1].key_value[VERSION_DISMISSED_KEY].value``` but ```_.get``` does ```null```/```undefined``` checks on the way.

---

*[ui/ts/models/cockroachlabs.ts, line 105 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-KDkkRN7VYvMauQuYB7G:-KDogX6oF9wKfKYAK1Lr:-534735270) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b1b6abd0a9e22d78032aa6c00d86469bbf38e9/ui/ts/models/cockroachlabs.ts#L105)):*
Whoops that's my local test server. Fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-:-KDohdoEa8DXZ1nDFMea:-1834867277)*

**Comment by cuongdo** [comment]
:lgtm: 

Took your branch for a quick spin, and looks good.

---

Reviewed 2 of 3 files at r5.
Review status: 17 of 19 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5569#-:-KDy9E99067SWHRepNEY:-1227887246)*

**Pull request merged**